### ISSUE TITLE
Grid Editor throws IllegalStateException

### ISSUE THREAD

**Issue by KochTobi** [vaadin/flow-components#6755]
### Description of the bug

When having a grid displaying java records, from 24.5.x on the `editItem` method throws an exception. This did not occur before as there is a custom renderer and it was not an issue before 24.5.x

Seems to have something to do on how the Binder for the editor is initialized.
```java
var binder = new Binder<>(MyClass.class);
```
versus
```java
Binder<MyClass> binder = new Binder<>();
```

```txt
java.lang.IllegalStateException: setBean can't be used with records, call readBean instead
	at com.vaadin.flow.data.binder.Binder.setBean(Binder.java:2360) ~[flow-data-24.5.2.jar:24.5.2]
	at com.vaadin.flow.component.grid.editor.EditorImpl.requestEditItem(EditorImpl.java:157) ~[vaadin-grid-flow-24.5.2.jar:na]
	at com.vaadin.flow.component.grid.editor.EditorImpl.lambda$editItem$2d5ea2ce$1(EditorImpl.java:140) ~[vaadin-grid-flow-24.5.2.jar:na]
	at com.vaadin.flow.internal.StateTree.lambda$runExecutionsBeforeClientResponse$2(StateTree.java:397) ~[flow-server-24.5.2.jar:24.5.2]
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:184) ~[na:na]
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:179) ~[na:na]
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1708) ~[na:na]
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509) ~[na:na]
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499) ~[na:na]
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151) ~[na:na]
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174) ~[na:na]
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[na:na]
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:596) ~[na:na]
	at com.vaadin.flow.internal.StateTree.runExecutionsBeforeClientResponse(StateTree.java:392) ~[flow-server-24.5.2.jar:24.5.2]
	at com.vaadin.flow.server.communication.UidlWriter.encodeChanges(UidlWriter.java:394) ~[flow-server-24.5.2.jar:24.5.2]
	at com.vaadin.flow.server.communication.UidlWriter.createUidl(UidlWriter.java:170) ~[flow-server-24.5.2.jar:24.5.2]
	at com.vaadin.flow.server.communication.UidlRequestHandler.createUidl(UidlRequestHandler.java:166) ~[flow-server-24.5.2.jar:24.5.2]
	at com.vaadin.flow.server.communication.UidlRequestHandler.writeUidl(UidlRequestHandler.java:156) ~[flow-server-24.5.2.jar:24.5.2]
	at com.vaadin.flow.server.communication.UidlRequestHandler.synchronizedHandleRequest(UidlRequestHandler.java:118) ~[flow-server-24.5.2.jar:24.5.2]
```

### Expected behavior

The editor component should show loaded with the information from the record (as was the case in 24.4.13) no matter how the binder is initialized.

### Minimal reproducible example

https://github.com/KochTobi/broken-vaadin-binder
You can find a working example at
https://github.com/KochTobi/broken-vaadin-binder/tree/working-example

### Versions

- Vaadin / Flow version: 24.5.2
- Java version: 21
- OS version: MacOs 14.7
- Browser version (if applicable): Firefox 131.0.3
- Application Server (if applicable):
- IDE (if applicable):

**Comment by KochTobi** [comment]
> I suppose a workaround could be to call `grid.getEditor().setBuffered(true)`

This workaround works :+1: Thank you!

However as this occurs at runtime I have to go through the code and see how binders are initialized. It would be great if there would be no need for this and however the Binder constructor is called, it should be compatible with the grid IMHO.

**Comment by vursen** [comment]
Using `setBuffered(true)` doesn't solve the issue. It just postpones the exception until `save()` is called. It seems that the ticket issue could be considered rather a request for adding records support to GridEditor. Records are immutable, so they already by design kind of act like mutable beans in buffered mode. This leads me to think that `save()` should also be updated to work with records. However, there are questions:

- should `save()` additionally support records, or should buffered mode be required to use records?
- should `closeEditor()` be allowed with records, or should `cancel()` be used instead?

**Comment by vursen** [comment]
We will fix this specific use case that worked in 24.4, and consider adding proper Java records support in buffered mode later:
- https://github.com/vaadin/flow-components/issues/6785
